src/main/kotlin/com/avail/utility/structures/EnumMap.kt

```diff
@@ -0,0 +1,210 @@
+ *
+ * @param K
+ *   The type of Enum keys maintained by this [EnumMap]
+ * @param V
+ *   The type of the values stored in this [EnumMap].
+ * @property enums
+ *   The [Array] of [Enum]s, in [Enum.ordinal] order of all the `Enum` values
+ *   represented in the `Enum`. It is expected that **all** enum values will
+ *   be present.
+ * @property values
+ *   The [Array] of values stored in this `Enum`. `null` values are not allowed
+ *   as a null value indicates the absence of a value.
+ *
+ * @constructor
+ * Construct an [EnumMap].
+ *
+ * @param enums
+ *   The [Array] of [Enum]s, in [Enum.ordinal] order of all the `Enum` values
+ *   represented in the `Enum`. It is expected that **all** enum values will
+ *   be present.
+ * @param sourceValues
+ *   The [Array] of values stored in this `Enum`.
+ */
+@Suppress("UNCHECKED_CAST", "unused")
+class EnumMap<K: Enum<K>, V : Any> constructor(
+	private val enums: Array<K>,
+	private val sourceValues: Array<V?>) : Map<K, V>
+{
+	init
+	{
+		assert(enums.size == sourceValues.size)
+	}
+
+	override val size: Int = sourceValues.count { it != null }
+
+	override val keys: Set<K> get() =
+		sourceValues.indices.filter { sourceValues[it] != null }
+			.map { enums[it] }.toMutableSet()
+
+	override val values: Collection<V>
+		get() = sourceValues.filterNotNull().toMutableList()
+
+	/**
+	 * Answer the value at the provided key. If no value is available (`null`
+	 * is at [Enum.ordinal] in the [sourceValues] array), evaluate the lambda,
+	 * place the resulting value in this map at that key, then answer the value.
+	 *
+	 * @param key
+	 *   The key to look up.
+	 * @param mappingFunction
+	 *   The function to run to produce a value if none found.
+	 * @return The stored value or the result of the [mappingFunction].
+	 */
+	fun getOrPut (key: K, mappingFunction: (K) -> V): V
+	{
+		var value = sourceValues[key.ordinal]
+		if (value == null)
+		{
+			value = mappingFunction(key)
+			sourceValues[key.ordinal] = value
+		}
+		return value
+	}
+
+	/**
+	 * Removes the specified key and its corresponding value from this
+	 * [EnumMap].
+	 *
+	 * @return
+	 *   The previous value associated with the key, or `null` if the key was
+	 *   not present in the map.
+	 */
+	fun remove(key: K): V?
+	{
+		val value = sourceValues[key.ordinal]
+		sourceValues[key.ordinal] = null
+		return value
+	}
+
+	/**
+	 * Removes all elements from this [EnumMap].
+	 */
+	fun clear()
+	{
+		for (i in sourceValues.indices)
+		{
+			sourceValues[i] = null
+		}
+	}
+
+	override fun containsKey(key: K): Boolean = true
+
+	override fun containsValue(value: V): Boolean =
+		sourceValues.any { value == it }
+
+	override fun get(key: K): V = sourceValues[key.ordinal]!!
+
+	override fun isEmpty(): Boolean = false
+
+	override val entries: Set<Map.Entry<K, V>>
+		get() = enums.map {
+			object : Map.Entry<K, V>
+			{
+				override val key: K = it
+				override val value: V = sourceValues[it.ordinal]!!
```

Comment:
I think we want to eliminate null values from the resulting set entirely.  Maybe:
```
get() = enums.map {
    object : Map.Entry<K, V>
    {
        override val key: K = it
        override val value: V = sourceValues[it.ordinal]     [i.e., no !!]
    }.filter { it.value != null }.toSet()
```

Reply:
Ended up using the code below instead as the proposed way broke this line:
`override val value: V = sourceValues[it.ordinal]` as value becomes V? which is not the expected override type

```
override val entries: Set<Map.Entry<K, V>>
		get() = enums.filter { sourceValues[it.ordinal] != null }.map {
			object : Map.Entry<K, V>
			{
				override val key: K = it
				override val value: V = sourceValues[it.ordinal]!!
			}
		}.toSet()
```